sections/tools/visual-regression-testing.md

```diff
@@ -0,0 +1,164 @@
+
+**The second selling point of services is how they can address cross viewport and browser concerns with a single test, so that we do not have to repeat the same test with different variants in CI**.
+
+
+
+### Cypress examples with Percy and Applitools
+
+> All the code samples can be found in the [full repo](https://github.com/muratkeremozcan/react-hooks-in-action-with-cypress).
+
+We will go through two popular services, Percy and Applitools, and showcase how services can save bandwidth with the maintenance of visual snapshots. We will also cover how they can be a force multiplier when we are concerned with cross browser and viewport combinations.
+
+Visual regression testing with services has a common flow:
+
+- Record a default snapshot and compare that default with the new, in subsequent test executions. We have to accept the initial snapshot once.
+- From then on, new snapshots matching the default get auto-accepted.
+
+- Non-matching new snapshots prompt a notification on the web interface; we either have to reject or accept this new baseline. If we reject, it is a defect. If we accept we have a new base line and the cycle continues.
+
+Suppose we want to verify the user's avatar. 
+
+![Image description](https://dev-to-uploads.s3.amazonaws.com/uploads/articles/rnl9mszbi1uuwr0ytgaf.png)
+
+#### Percy flow:
+
+* Here is [a simple test](https://github.com/muratkeremozcan/react-hooks-in-action-with-cypress/blob/main/cypress/e2e/ui-integration/user-context-retainment.spec.js) that verifies the avatar.
+
+
+```javascript
+// cypress/e2e/ui-integration/user-context-retainment.spec.js
+describe('User selection retainment between routes',
+  { tags: '@ui-integration' },
+  () => {
+    before(() => {
+      cy.stubNetwork()
+      cy.visit('/')
+    })
+
+    it('Should keep the user context between routes', () => {
+      cy.fixture('users').then((users) => {
+        cy.get('.user-picker').select(users[3].name)
+        cy.contains('Users').click()
+
+        cy.wait('@userStub')
+        cy.url().should('contain', '/users')
+        cy.get('.item-header').contains(users[3].name)
+
+        // the visual test
+        cy.percySnapshot('User selection retainment between routes')
+
+        // using custom command for css selector focus
+        cy.getByCy('user-details').percySnapshotElement(
+          'user details with custom selector'
+        )
+      })
+    })
+  }
+)
+```
+
+We see the first one-liner taking a full screen shot, and the subsequent call taking a snapshot of the `user-details` selector. Selector focused snapshots are not natively supported in Percy, therefore [a custom command](https://github.com/muratkeremozcan/react-hooks-in-action-with-cypress/blob/main/cypress/support/commands.js#L112) `percySnapshotElement` is used.
```

Comment:
**Suggestion**: Adding the custom command above the test or splitting the tests and explanations into two different parts: the first is a classic snapshot, the second is a selector-based snapshot due to Percy not supporting it, etc.

Reply:
Made 2 tests.